### PR TITLE
Create hex image

### PR DESCRIPTION
**Pull request by alekob**
Create an extension to "generatebits" in order to write a hex image data and / or instruction memories. This enables the user to use one single (e. g. ROM) image for ASIC simulation and FPGA prototyping.

### CODE REVIEW

File: tce/src/applibs/PIG/AsciiImageWriter.cc [hunk 1]
```diff
@@ -138,3 +144,62 @@ AsciiImageWriter::writeSequence(
...
+ *                       the row.
+ */
+void
+AsciiImageWriter::writeHexSequence(
+    std::ostream& stream, int length, bool padEnd) const
+    throw (OutOfRange) {
+    int lastIndex = nextBitIndex_ + length - 1;
+
+    if (lastIndex >= bits_.size() && !padEnd) {
+        const string procName = "AsciiImageWriter::writeHexSequence";
```
Comment (pjaaskel) [inline tce/src/applibs/PIG/AsciiImageWriter.cc]: You can replace this with \_\_func\_\_.

File: tce/src/applibs/PIG/AsciiImageWriter.cc [hunk 2]
```diff
@@ -138,3 +144,62 @@ AsciiImageWriter::writeSequence(
...
+        throw OutOfRange(__FILE__, __LINE__, procName);
+    }
+
+    const int nibbleCount = static_cast<int>(
+        ceil(static_cast<double>(length / 4)));
+
+    std::vector<bool> bitRow;
+
+    // Copy bitstream for one row
+    for(int i=nextBitIndex_; i<=lastIndex; ++i) {
```
Comment (pjaaskel) [inline tce/src/applibs/PIG/AsciiImageWriter.cc]: Plase add space after 'for'

File: tce/src/applibs/PIG/AsciiImageWriter.cc [hunk 3]
```diff
@@ -138,3 +144,62 @@ AsciiImageWriter::writeSequence(
...
+        bitRow.push_back(bits_[i]);
+    }
+
+    // optionally extend bit stream
+    int numPadBits = 4*nibbleCount - length;
```
Comment (pjaaskel) [inline tce/src/applibs/PIG/AsciiImageWriter.cc]: Please add space after and before *

File: tce/src/applibs/PIG/AsciiImageWriter.cc [hunk 4]
```diff
@@ -138,3 +144,62 @@ AsciiImageWriter::writeSequence(
...
+    if (numPadBits) {
+        std::vector<bool> padBits(numPadBits, 0);
+        bitRow.insert(bitRow.begin(), padBits.begin(), padBits.end());
+    }
+
+    // Generate a list of nibble
+    std::vector<uint8_t> Nibble;
+    for (std::vector<bool>::iterator it=bitRow.begin(); it<bitRow.end(); it+=4) {
```
Comment (pjaaskel) [inline tce/src/applibs/PIG/AsciiImageWriter.cc]: Ditto before/after =.

File: tce/src/applibs/PIG/AsciiImageWriter.cc [hunk 5]
```diff
@@ -138,3 +144,62 @@ AsciiImageWriter::writeSequence(
...
+        Nibble.push_back(
+            std::accumulate(
+                it, it+4, 0, [] (int x, int y) { return (x << 1) + y; }
+            )
+        );
+    }
+
+    // "print" hex stream
+    for (std::vector<uint8_t>::iterator ui8=Nibble.begin(); ui8<Nibble.end(); ++ui8) {
```
Comment (pjaaskel) [inline tce/src/applibs/PIG/AsciiImageWriter.cc]: White spaces needed here. Also check that lines are below 80 chars of length.

File: tce/src/applibs/PIG/AsciiImageWriter.hh [hunk 6]
```diff
@@ -34,6 +34,9 @@
 #ifndef TTA_ASCII_IMAGE_WRITER_HH
 #define TTA_ASCII_IMAGE_WRITER_HH
 
+#include <cstdint>
```
Comment (pjaaskel) [inline tce/src/applibs/PIG/AsciiImageWriter.hh]: If not needed in the header, please move the includes to the .cc to reduce cpp effort.

File: tce/src/applibs/PIG/HexImageWriter.cc [hunk 7]
```diff
@@ -0,0 +1,90 @@
+/*
+    Copyright (c) 2002-2018 Tampere University of Technology.
```
Comment (pjaaskel) [inline tce/src/applibs/PIG/HexImageWriter.cc]: Also add your name to the copyright as you own the code (no copyright assignments to TUT to my knowledge). Tampere University of Technology and [YOURNAME] is probably fair as it uses TUT written interfaces.

File: tce/src/applibs/PIG/HexImageWriter.hh [hunk 8]
```diff
@@ -0,0 +1,52 @@
+/*
+    Copyright (c) 2002-2018 Tampere University of Technology.
+
+    This file is part of TTA-Based Codesign Environment (TCE).
```
Comment (pjaaskel) [inline tce/src/applibs/PIG/HexImageWriter.hh]: Here too.

**Comment by pjaaskel** [comment]
Info for developers: http://tce.cs.tut.fi/cgi-bin/moin.cgi/Development the testing guide: http://tce.cs.tut.fi/cgi-bin/moin.cgi/TestSuite

**Comment by tervoa** [comment]
Program image generation with an instruction width not divisible by 4 (e.g. `minimal.adf`) fails with the following:
```
terminate called after throwing an instance of 'std::length_error'
  what():  vector<bool>::_M_insert_range
Aborted
```

**Comment by alekob** [comment]
Thanks for all of your helpful comments and testings. (Hopefully) I've fixed all issues and added a new test in my latest commit a2c4fb9802e3383d4bfb5632ec4cd182180abea2

**Pull request merged**